packages/flutter/lib/src/widgets/restoration.dart

```diff
@@ -0,0 +1,1002 @@
+      child: widget.child,
+    );
+  }
+}
+
+/// Inserts a provided [RestorationBucket] into the widget tree and makes it
+/// available to descendants via [RestorationScope.of].
+///
+/// {@macro flutter.widgets.restoration.scope}
+///
+/// If [bucket] is null, no restoration bucket is made available to descendant
+/// widgets ([RestorationScope.of] invoked from a descendant will return null).
+/// This effectively turns off state restoration for the subtree because no
+/// bucket for storing restoration data is made available.
+///
+/// See also:
+///
+///  * [RestorationScope], which inserts a bucket obtained from a surrounding
+///    restoration scope into the widget tree and makes it accessible
+///    for descendants via [RestorationScope.of].
+///  * [RootRestorationScope], which inserts the root bucket provided by
+///    the [RestorationManager] into the widget tree and makes it accessible
+///    for descendants via [RestorationScope.of].
+///  * [RestorationMixin], which may be used in [State] objects to manage the
+///    restoration data of a [StatefulWidget] instead of manually interacting
+///    with [RestorationScope]s and [RestorationBucket]s.
+///  * [RestorationManager], which describes the basic concepts of state
+///    restoration in Flutter.
+class UnmanagedRestorationScope extends InheritedWidget {
+  /// Creates an [UnmanagedRestorationScope].
+  const UnmanagedRestorationScope({
+    Key key,
+    this.bucket,
+    Widget child,
+  }) : super(key: key, child: child);
+
+  /// The [RestorationBucket] that this widget will insert into the widget tree.
+  ///
+  /// Descendant widgets may obtain this bucket via [RestorationScope.of].
+  final RestorationBucket bucket;
+
+  @override
+  bool updateShouldNotify(UnmanagedRestorationScope oldWidget) {
+    return oldWidget.bucket != bucket;
+  }
+}
+
+/// Inserts a child bucket of [RestorationManager.rootBucket] into the widget
+/// tree and makes it available to descendants via [RestorationScope.of].
+///
+/// {@macro flutter.widgets.restoration.scope}
+///
+/// The exact behavior of this widget depends on its ancestors: When the
+/// [RootRestorationScope] does not find an ancestor restoration bucket via
+/// [RestorationScope.of] it will claim a child bucket from the root restoration
+/// bucket ([RestorationManager.rootBucket]) using the provided [restorationId]
+/// and inserts that bucket into the widget tree where descendants may access it
+/// via [RestorationScope.of]. If the [RootRestorationScope] finds a non-null
+/// ancestor restoration bucket via [RestorationScope.of] it will behave like a
+/// regular [RestorationScope] instead: It will claim a child bucket from that
+/// ancestor and insert that child into the widget tree.
+///
+/// Unlike the [RestorationScope] widget, the [RootRestorationScope] will
+/// grantee that descendants have a bucket available for storing restoration
+/// data as long as [restorationId] is not null.
```

Comment:
From reading this and the previous doc, it's not clear to me why I'd ever not use this widget.

Alternatively, could we just simplify these into a single widget that takes a boolean about whether it should create a root bucket if needed or not?

Reply:
not use this widget -> instead of the other RestorationScope widget.

Reply:
If you implement a stand-alone widget, you want to use a regular restoration scope to create namespaces for restoration ids. The widget can then be used in applications that want to do state restoration and those that don't. If the regular restoration scope doesn't find an ancestor scope, it will not cause any information to be contributed to the restoration data. In fact, the entire logic of retrieving the restoration data from the engine, parsing it, encoding the restoration data when it changes, sending it to the engine will not trigger.

If you'd use a RootRestorationScope, even in an application that doesn't want to do state restoration all that logic now triggers for the part of the app that's wrapped in this thing.

So that's the biggest difference: RootRestorationScope essentially forces you into doing state restoration. The regular scope doesn't have any opinion on whether you actually want to store restoration data or not.

The root restoration scope should only be used at the very top of the app (in fact, I am going to integrate it into WidgetApp).

Let me update the docs on this a little to make that clearer.

> Alternatively, could we just simplify these into a single widget that takes a boolean about whether it should create a root bucket if needed or not?

Yes, we could just have this implemented in one widget with a flag. The implementation of that widget would be pretty complex though. And I feel that the difference in functionality is more than just a config flag. It is a whole different widget.